src/server/mod.rs

```diff
@@ -172,21 +186,67 @@ where
 
 // ===== impl Connection =====
 
-impl<T, S, E, B, F> Connection<T, S, E, B, F>
+impl<T, S, E, B, F> Future for Connection<T, S, E, B, F>
 where T: AsyncRead + AsyncWrite,
       S: NewService<Request = http::Request<RecvBody>, Response = Response<B>>,
-      B: Body,
+      E: Executor<Background<<S::Service as Service>::Future, B>>,
+      B: Body + 'static,
+      F: Modify,
 {
-    fn is_ready(&self) -> bool {
-        use self::State::*;
+    type Item = ();
+    type Error = Error<S>;
 
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let ret = (|| loop {
```

Comment:
This took a moment to unpack. I'd prefer a separate fn but I won't block the PR on it.